src/robot/running/keywordrunner.py

```diff
@@ -206,16 +249,69 @@ def _to_number_with_arithmetics(self, item):
             raise TypeError("Expected number, got %s." % type_name(item))
         return number
 
-    def _run_one_round(self, data, values):
-        name = ', '.join(format_assign_message(var, item)
-                         for var, item in zip(data.variables, values))
-        result = KeywordResult(kwname=name,
-                               type=data.FOR_ITEM_TYPE)
-        for var, value in zip(data.variables, values):
-            self._context.variables[var] = value
-        runner = KeywordRunner(self._context, self._templated)
-        with StatusReporter(self._context, result):
-            runner.run_keywords(data.keywords)
+
+class ForInZipLoopRunner(BasicForLoopRunner):
+    def __init__(self, forstep, templated=False):
+        super(ForInZipLoopRunner, self).__init__(forstep, templated)
+
+    def _flavor_name(self):
+        return 'IN ZIP'
+
+    def _transform_items(self, items):
+        answer = list()
+        for item in items:
+            if not is_list_like(item):
+                # Checking here because _validate() doesn't have access
+                # to the actual values.
+                raise DataError("For-In-Zip Loop items must all be List-like (but not Strings); got %s with value '%s'" % (item.__class__, item))
+        for zipped_item in zip(*items):
+            answer.extend(zipped_item)
+        return answer
+
+class ForInEnumerateLoopRunner(BasicForLoopRunner):
+    def __init__(self, forstep, templated=False):
+        super(ForInEnumerateLoopRunner, self).__init__(forstep, templated)
+
+    def _flavor_name(self):
+        return 'IN ENUMERATE'
+
+    def _transform_items(self, items):
+        answer = list()
+        for zipped_item in zip(*items):
+            answer.extend(zipped_item)
+        return answer
+
+    @property
+    def _original_values_per_iteration(self, values):
+        """
+        The number of values per iteration;
+        used to check if we have (a multiple of this) values.
+
+        This is its own method to support loops like ForInEnumerate
+        which add/remove items to the pool.
+        """
+        if len(values) < 2:
+            raise DataError('FOR IN ENUMERATE expected 2+ loop variables, got %d.'
+                    % len(values))
+        return len(values) - 1
+
+    def _get_values_for_one_round(self, data):
+        for idx, values in enumerate(super(ForInEnumerateLoopRunner, self)
+                ._get_values_for_one_round(data)):
+            yield (idx,) + values
+
+
+# Defining this dict here in for potentially easier extension by libraries...
+# ...unless that's a terrible idea?
+_FOR_LOOP_FLAVORS = dict(
+        INRANGE=ForInRangeLoopRunner,
+        INZIP=ForInZipLoopRunner,
+        INENUMERATE=ForInEnumerateLoopRunner,
+        )
```

Comment:
Do you have use cases for libraries or other code adding new for loops? That sounds a bit weird but I'm definitely not interested in preventing it. On the other hand, we cannot generally promise APIs like this are stable between major releases. The only way to make sure APIs don't break is adding unit or acceptance tests, but even then keeping API compatibility might be a burden.